### PR TITLE
api: Discourage using ClientInterceptor for credentials

### PR DESCRIPTION
**Pull request by ejona86**
As brought up in #5665

### CODE REVIEW

File: api/src/main/java/io/grpc/ClientInterceptor.java [hunk 1]
```diff
@@ -24,10 +24,14 @@
...
  * <ul>
- * <li>Adding credentials to header metadata</li>
  * <li>Logging and monitoring call behavior</li>
+ * <li>Adding metadata for proxies to observe</li>
  * <li>Request and response rewriting</li>
  * </ul>
+ *
+ * <p>Providing authentication credentials is better served by {@link
+ * CallCredentials}. But a {@code ClientInterceptor} could set the {@code
+ * CallCredentials} within the {@link CallOptions}.
```
Comment (sanjaypujare) [inline api/src/main/java/io/grpc/ClientInterceptor.java]: Should we say something about whether this is undesirable or not?

File: api/src/main/java/io/grpc/ClientInterceptor.java [hunk 2]
```diff
@@ -24,10 +24,14 @@
  * <p>Implementers use this mechanism to add cross-cutting behavior to {@link Channel} and
  * stub implementations. Common examples of such behavior include:
  * <ul>
- * <li>Adding credentials to header metadata</li>
  * <li>Logging and monitoring call behavior</li>
+ * <li>Adding metadata for proxies to observe</li>
```
Comment (sanjaypujare) [inline api/src/main/java/io/grpc/ClientInterceptor.java]: Metadata can be added for anything - why limit it to "for proxies to observe"?  (I know it's only an example but still...)

File: api/src/main/java/io/grpc/ClientInterceptor.java (hunk 1 above)
Comment (ejona86) [reply api/src/main/java/io/grpc/ClientInterceptor.java]: ClientInterceptor can work. But using CallCredentials is easier and generally better. It's sort of in-depth, so didn't want to really get into it in this documentation.

We could end up putting some benefits of CallCredentials in the CallCredentials class though.

File: api/src/main/java/io/grpc/ClientInterceptor.java (hunk 2 above)
Comment (ejona86) [reply api/src/main/java/io/grpc/ClientInterceptor.java]: These are just examples. There are other possibilities. The listed items are use-cases, not just functionality that is exposed. When someone starts reading this class, I wouldn't expect them to know what metadata is and why it is used.

**Pull request merged**